scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1089,25 +1089,72 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
         case (pipe, cnt) => pipe.flatMap(List.fill(cnt)(_).iterator)
       }
       .map(_.toPipe[U](fieldNames)(flowDef, mode, setter)) // linter:ignore
-      .toList
+      .toVector
 
     if (merged.size == 1) {
       // there is no actual merging here, no need to rename:
       merged.head
     } else {
-      merged.reduce[Pipe] {
-        case (left, right) =>
-          // special handling for cases where one side of the hashjoin is merged
-          // with the hashjoin result. Cascading no longer allows it,
-          // so we add a checkpoint to the join result as a workaround
-          if (RichPipe.isHashJoinedWithPipe(left, right))
-            new Merge(RichPipe.assignName(new Checkpoint(left)), RichPipe.assignName(right))
-          else if (RichPipe.isHashJoinedWithPipe(right, left))
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(new Checkpoint(right)))
-          else
-            new Merge(RichPipe.assignName(left), RichPipe.assignName(right))
-      }
+      val prepared  = mergeAvoidingHashes(merged)
+      prepared
+    }
+  }
+
+  /**
+    * Special handling for cases where one side of the hashjoin is merged
+    *  with the hashjoin result. Cascading no longer allows it (as of 3.0),
+    *  so we insert checkpoints and/or intermediate merge stages as appropriate
+    *
+    * @param pipes a list of pipes possibly colliding with each other
+    * @return an updated list of pipes, which can safely be merged
+    */
+  protected def mergeAvoidingHashes(pipes: Vector[Pipe]): Pipe = {
+    require(!pipes.isEmpty)
+
+    // we make use of the fact that the pipe merge operation is not just associative but also commutative
+    val (colliding, uncolliding) = pipes.partition(p => pipes.find(o => (o != p) && RichPipe.isHashJoinedWithPipe(p, o)).isDefined)
```

Comment:
so, basically we are building a directed graph: `a -> b` if `a isHashJoinedWith b`. Then pipes with no edges are in the uncolliding set.

Then, we need to deal with the colliding set. For that, we just checkpoint all the colliding nodes, then do one big merge.

Is this correct? If so, why do the recursion? It seems we could just do:
```scala
new Merge(avoidNameClashes(colliding).map(new Checkpoint(_)) ++ uncolliding: _*)
```
I don't quite see why we need the recursion in the case.